### PR TITLE
add Sliding Window Attention support to cudnn flash attn api

### PR DESCRIPTION
**Pull request by kocchop**
# Description

Start with a short description of what the PR does and how this is a change from the past.

This PR enables usage of Sliding Window Attention with cudnn flash attn api. Sliding window attention only works for causal masking for now. Hence, inserted necessary checks to choose correct mask type. TE only accepts generated mask when `mask_type="padding_causal"` but this is not supported with SWA yet. With the PR, we can now run the class of Gemma models with proper GPU flash attention now.

If the change fixes a bug or a Github issue, please include a link, e.g.,:

# Tests

I ran gemma2-9b model which uses sliding window attn and head_dim=256; also tested llama2-7b model to make sure the existing path doesn't break

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/layers/attentions.py [hunk 1]
```diff
@@ -393,32 +393,42 @@ def cudnn_flash_attention(
...
+    2. Head_dim = 256 is also supported from TE-1.12 stable release
     """
     # These imports are only meant to work in a GPU build.
     from transformer_engine.jax.flax.transformer import DotProductAttention  # pytype: disable=import-error
 
     _, _, _, head_dim = query.shape  # pylint: disable=unused-variable
 
-    # generate attn_mask
-    attn_mask = self.generate_attention_mask(query, key, decoder_segment_ids, model_mode)
+    sliding_window_size = self.sliding_window_size
```
Comment (wang2yn84) [inline MaxText/layers/attentions.py:404]: Nit ,better assign sliding_window_size to None when for non sliding window case to better align with the requirement of the API, wdyt?

File: MaxText/layers/attentions.py [hunk 2]
```diff
@@ -393,32 +393,42 @@ def cudnn_flash_attention(
...
         dropout_rng_name="aqt",
         dtype=self.dtype,
         float32_logits=self.float32_logits,
         qkv_layout="BSHD_BSHD_BSHD",  # 'BS3HD', 'BSHD_BS2HD' or 'BSHD_BSHD_BSHD'
         scale_factor=1.0 / math.sqrt(head_dim),
         transpose_batch_sequence=False,
+        window_size=sliding_window_size,
     )
-    return dpa_layer(query, key, value, mask=attn_mask)
+    if self.attention_type == AttentionType.LOCAL_SLIDING:
```
Comment (wang2yn84) [inline MaxText/layers/attentions.py:428]: nit: just to keep the code clean if we set mask to either None or attn_mask, wdyt?

File: MaxText/layers/attentions.py (hunk 1 above)
Comment (kocchop) [reply MaxText/layers/attentions.py:404]: sure, will change to None

**Comment by kocchop** [comment]
@wang2yn84 PTAL, I made the suggested change

**Comment by wang2yn84** [comment]
> @wang2yn84 PTAL, I made the suggested change

Great! Thank you! Lgtm.

**Review by wang2yn84 — APPROVED** [review]

**Review by gobbleturk — APPROVED** [review]
